Initial Beta System

annopkdd opened the pull request:
**To do**
- [x] Receiver
- [x] Command
- [x] Command Register
- [x] Command Bus
- [x] Middleware
- [x] Handle
- [x] Message Template/Message Factory

Commit:
initial system

initial all base system

Commit:
delete ACCESS_TOKEN and SECRET

Commit:
fix read me

Commit:
delete requirements file

Commit:
fix read me2

liverbool commented:
@criticalnooks Can you port this php function to python?

```php
    /**
     * @param Middleware[] $middlewareList
     *
     * @return callable
     */
    function createExecutionChain(array $middlewareList)
    {
        $lastCallable = function () {
            // the final callable is a no-op
        };

        while ($middleware = array_pop($middlewareList)) {
            $lastCallable = function ($command) use ($middleware, $lastCallable) {
                return $middleware->execute($command, $lastCallable);
            };
        }

        return $lastCallable;
    }
```

Commit:
core-python V2

MiddleWare
- button_template_middleware
- carousel_template_middleware
- confirm_template_middleware
- default_message_middleware
- image_carousel_template_middleware
- image_message_middleware
-
 imagemap_send_message_middleware
- location_message_middleware
- sticker_message_middleware

Structure
- use dependency injection for middleware

Commit:
fix bug1

Commit:
fix bug2

Commit:
fix bug3

Commit:
fix bug

Commit:
fix bug

SyntaxError: invalid syntax

Commit:
inherit middleware class

Commit:
fix bug

TypeError: execute() takes 1 positional argument but 2 were given

Commit:
fix bug

NameError: name 'Command' is not defined

Commit:
fix bug

 AttributeError: 'MessageTemplateFactory' object has no attribute 'template'

Commit:
fix bug

 KeyError: 'longtitude'

Commit:
fix bug

KeyError: 'alt_text'

Commit:
fix bug

 Process exited with status 3

Commit:
fix bug

CarouselTemplate
ImagemapSendTemplate

Commit:
fix bug

 KeyError: 'template'

Commit:
fix bug

KeyError: 'template'(again)

Commit:
fix bug

 NameError: name 'LocationSendMessage' is not defined

Commit:
fix content

change base_url ImageMapSendMessage

Commit:
fix content

ImageMapSendMessage
change base_url
change link_uri

Commit:
Take out ImagemapMessage

Commit:
fix content

DefaultMessageMiddleware
fix word

liverbool commented:
👍 looks good!

I thinks we need to small change to something like this (?)

### pseudo code
```python
class DefaultCommand():
    property name = 'DefaultCommand'
    property cmd = ''
    property template = None

    private event;

    self. __init__(event):
        self.event = event;

    self.isValidCmd():
        return self.cmd == self.event.input

class CommandA():
    property name = 'CommandA'
    property cmd = '/a'
    property template = None

    private event;

    self. __init__(event):
        self.event = event;

    self.isValidCmd():
        return self.cmd == self.event.input

class CommandB():
    property name = 'CommandB'
    property cmd = '/b'
    property template = None

    private event;

    self. __init__(event):
        self.event = event;

    self.isValidCmd():
        return self.cmd == self.event.input

class CommandRegistry():
    private commands = []

    def add(command):
        self.commands.push(command)

    def find(event):
        for (self.commands as command) {
            command = new command(event);
            if (command.isValidCmd()) {
                return command;
            }
        }

        return new DefaultCommand(event);

# Usage
commandRegistry = new CommandRegistry();
commandRegistry.add(CommandA)
commandRegistry.add(CommandB)

# middleware
class MiddlewareX():
    def execute(command):
        if (command.name != 'CommandA') {
            return;
        }

        # i want to tell this command to send vdo message
        command.template = new VdoMessageTemplate(....);

```

### Command should have Abstract -- to handle common logic of commands.
```python
from abc import ABCMeta, abstractmethod

class Command:
    __metaclass__ = ABCMeta
    
    property event;
    property template;

    def setEvent(event):
        self.event = event;

    @abstractmethod
    def isValidCmd(self): pass

class CommandA(Command):
    property cmd
    property name

    def isValidCmd():
        return self.cmd === self.event.input

# in CommandRegistry.find(event) 
....
    command = new CommandA()
    command.setEvent(event)
....
```

Commit:
core-python V2.1

- modify command object

Commit:
fix bug

AttributeError: 'CommandRegistry' object has no attribute 'check'

Commit:
fix bug

NameError: name 'self' is not defined

Commit:
fix bug

AttributeError: 'Command1' object has no attribute 'isValidCmd'

Commit:
fix bug

NameError: name 'default_command' is not defined

Commit:
debug

AttributeError: 'NoneType' object has no attribute 'get_command'

Commit:
debug

AttributeError: 'NoneType' object has no attribute 'get_command'

Commit:
fix bug

AttributeError: 'NoneType' object has no attribute 'get_command'

Commit:
debug

AttributeError: 'NoneType' object has no attribute 'get_command'

Commit:
debug

Commit:
fix bug

I think it's work now!

Commit:
debug

: TypeError: string indices must be integers

Commit:
fix bug

TabError: inconsistent use of tabs and spaces in indentation

Commit:
fix bug

SyntaxError: invalid syntax

Commit:
debug

debug hello middleware

Commit:
debug

Commit:
debug

Commit:
debug

Commit:
debug

Commit:
delete pycache

Commit:
debug

Commit:
fix bug debug

AttributeError: 'HelloMiddleware' object has no attribute 'command'

Commit:
debug

in all middleware

Commit:
debug

Commit:
debug

Commit:
fix bug

Commit:
remove test file
